tests/keithley_2700.py

```diff
@@ -0,0 +1,188 @@
+import unittest
+from unittest import skipIf
+from utils.channel_access import ChannelAccess
+from utils.testing import get_running_lewis_and_ioc
+from utils.ioc_launcher import IOCRegister
+
+
+
+class Keithley_2700Tests(unittest.TestCase):
+    """
+    Tests for the Keithley2700.
+    """
+
+    def setUp(self):
+        self._lewis, self._ioc = get_running_lewis_and_ioc("keithley_2700")
+        self.ca = ChannelAccess(default_timeout=30)
+        self.ca.wait_for("KHLY2700_01:IDN")
+
+    @skipIf(IOCRegister.uses_rec_sim, "In rec sim this test fails")
+    def test_GIVEN_idn_defined_WHEN_read_THEN_idn_is_as_expected(self):
+        self.ca.assert_that_pv_is("KHLY2700_01:IDN", "Keithley 2700 Multimeter emulator.")
+
+    @skipIf(IOCRegister.uses_rec_sim, "In rec sim this test fails")
+    def test_GIVEN_available_AND_reserved_bytes_set_WHEN_read_THEN_staTus_of_memory_is_as_expected(self):
+        self.ca.set_pv_value("KHLY2700_01:BUFFER_SIZE:SP", 100)
+        self._lewis.backdoor_set_on_device("buffer_readings", 10)
+        self._lewis.backdoor_set_on_device("buffer", 100)
+        self.ca.assert_that_pv_is("KHLY2700_01:BUFFER_SIZE.", 100)
+        self.ca.set_pv_value("KHLY2700_01:BUFFER_STATS.PROC", 1)
+        self.ca.assert_that_pv_is("KHLY2700_01:BUFFER_STATS", "92160 bytes,10240 bytes")
+
+    def test_GIVEN_source_set_WHEN_read_THEN_source_is_as_expected(self):
+        sample_data = [0, 1, 2, 3, 4]
+        expected_channel = ["IMM", "EXT", "TIM", "MAN", "BUS"]
+        for input, output in zip(sample_data, expected_channel):
```

Comment:
Zipping together two lists is risky. As the data isn't in itself coupled, you are open to the possibility of them changing independently and invalidating the tests. You could instead do this as:

```
expected_channels = {0: "IMM", 1: "EXT"...}
for channel_number in expected_channels.keys():
     ...
```

Alternatively, given the form of `sample_data` you could do:

```
for channel_number, channel_name in enumerate(expected_channel):
   ...
```

Please pick an improved pattern and apply it to all other instances where you've used `zip`